### PR TITLE
test: finalize environment setup and runtime for #415.

### PR DESCRIPTION
**Pull request by htuch**
This patch provides the missing environment setup pieces to support a complete run of all tests
under Bazel:

* Use a dedicated, dynamically created, temporary directory for Unix Domain Socket paths. This
  overcomes the sun_path limit in sockaddr_un (108 characters on Linux). Bazel generates very long (>
  150 chars in my case) paths for TEST_TMPDIR.

* Replace the per-test system(3) calls (e.g. in runtime_impl_test) with a --envoy_test_setup flag
  that test/main.cc takes, allowing a shell script to execute in the test environment to perform some
  setup.

* Some rejiggering of the JSON template substitution scripts - the current form (in
  environment_sub.sh) provides for a clean Bazel expression, at the expense of some hacks in
  run_envoy_tests.sh that will soon

Verified on both cmake and Bazel (//test/...).

### CODE REVIEW

File: test/main.cc [hunk 1]
```diff
@@ -25,6 +25,32 @@ int main(int argc, char** argv) {
   ::testing::Test::RecordProperty("TemporaryDirectory", TestEnvironment::temporaryDirectory());
   ::testing::Test::RecordProperty("RunfilesDirectory", TestEnvironment::runfilesDirectory());
 
+  if (::setenv("TEST_UDSDIR", TestEnvironment::unixDomainSocketDirectory().c_str(), 1) != 0) {
+    ::perror("Failed to temporary UDS directory.");
```
Comment (mattklein123) [inline test/main.cc]: typo

File: test/main.cc [hunk 2]
```diff
@@ -25,6 +25,32 @@ int main(int argc, char** argv) {
...
+    ::exit(1);
+  }
+
+  // Quick and dirty filtering and execution of --envoy_test_setup. Scripts run
+  // in this environment have access to env vars such as TEST_TMPDIR. This
+  // script can be used for situations where genrules are unsuitable for unit
+  // tests, since genrules don't have access to TEST_TMPDIR.
```
Comment (mattklein123) [inline test/main.cc]: I understand you are trying to get this done, but it would be nice if there was handled as part of a TestOptionsImpl below. Perhaps we can fix this when we convert to gflags. Can you put a TODO for @rlazarus  in?

File: test/run_envoy_tests.sh [hunk 3]
```diff
@@ -20,15 +20,33 @@ fi
...
-$SOURCE_DIR/test/config/integration/gen_test_configs.sh $SOURCE_DIR/test/config/integration \
-  $TEST_SRCDIR/test/config/integration
+
+# Some hacks for the config file template substitution. These go away in the Bazel build.
+CONFIG_IN_DIR="$SOURCE_DIR"/test/config/integration
+CONFIG_RUNFILES_DIR="$TEST_SRCDIR/$TEST_WORKSPACE"/test/config/integration
+CONFIG_OUT_DIR="$TEST_TMPDIR"/test/config/integration
+mkdir -p "$CONFIG_RUNFILES_DIR"
+mkdir -p "$CONFIG_OUT_DIR"
+cp "$CONFIG_IN_DIR"/*.json "$CONFIG_RUNFILES_DIR"
```
Comment (mattklein123) [inline test/run_envoy_tests.sh]: What is the point of copying any of this into runfiles first? What is the difference between runfiles and tmp? Sorry this is kind of hard to follow.

Comment (htuch) [inline test/run_envoy_tests.sh]: Under Bazel, runfiles are where the source dependencies and pregenerated (from other rules) dependencies live. They are read-only in the build/test sandbox. The tmpdir is where we can do writable stuff.

Unfortunately, we can't pregenerate the JSON configs for the tests and put them in runfiles, since they may have template variables that refer to tmpdir, which is only known at test execution time (not build time). So, we need to put them in tmpdir. 

The Bazel `environment_sub.sh` rule will take the JSON templates from runfiles and generate in tmpdir (see #642). With the cmake build, we actually need to copy from source to runfiles first to make this transparent to scripts like `environment_sub.sh`. This disappears soon when we remove cmake.
